### ISSUE TITLE
Still massive hitching

### ISSUE THREAD

**Issue by SteveBaggins** [esx-framework/esx_status#52]
I am still getting massive hitch times even after the new pulls and updates. Below is what I get, over time my server climbs 40-60 players and it just has massive hitching every like 10 seconds after that from the update function shown below. Its not big hitches, but its a lot and they pile up.

![esxstatus](https://user-images.githubusercontent.com/59070563/107138112-6320f780-68e0-11eb-868b-07b6223c1a6f.PNG)

**Comment by Tinky124** [comment]
@SteveBaggins Try to change `Config.UpdateInterval = 10000` to  `Config.UpdateInterval = 15000` to maybe lower that massive hitch time.. but idk how to fix that.. esx_status is bad...

**Comment by Tinky124** [comment]
https://github.com/esx-framework/esx_status/pull/53

**Comment by SteveBaggins** [comment]
> @SteveBaggins Try to change `Config.UpdateInterval = 10000` to `Config.UpdateInterval = 15000` to maybe lower that massive hitch time.. but idk how to fix that.. esx_status is bad...

Agreed, but is there any other options? I put in your pull, we'll see I had already had it at 20secs in that screenshot.

**Comment by SteveBaggins** [comment]
> #53

Still seems to be hitching heavily. I notice like if the server reboots and has 50 players, no hitches, its like over time as the server is running it just builds up over time. Eventually it just hitches like mad and the server memory usage goes up super high until it crashes the server eventually.

**Comment by yolobrandon01** [comment]
i also have this same issue (i think) where everthing is running fine for a few hours and then esx_status is causing the server it hitch incredibly bad over time, until the server restarts.

**Comment by yolobrandon01** [comment]
![image](https://user-images.githubusercontent.com/66453444/114635054-e3bdf480-9c91-11eb-8f89-4efe32de447d.png)

**Comment by Tinky124** [comment]
@yolobrandon01 Update to lastest esx_status version and also change time update in config file.

Try to change `Config.UpdateInterval = 10000` to `Config.UpdateInterval = 60000`

**Comment by yolobrandon01** [comment]
![image](https://user-images.githubusercontent.com/66453444/117753798-1999be80-b1e7-11eb-8601-613cfe802ebe.png)
Still having high hitches.. On ESX version 1.2FINAL

![image](https://user-images.githubusercontent.com/66453444/117753863-36ce8d00-b1e7-11eb-923a-4495bd1c4084.png)

**Comment by thelindat** [comment]
This problem should be resolved with #54, but only if you are using **ESX Legacy** (not 1.2, not v1 Final).

**Comment by yolobrandon01** [comment]
So the recent merge only fixes the Hitching for Legacy 1.1? Not 1.2 / v1Final?

**Comment by thelindat** [comment]
> 
> 
> So the recent merge only fixes the Hitching for Legacy 1.1? Not 1.2 / v1Final?

Legacy is the latest updates to ESX1 to solve issues such as this.
https://github.com/esx-framework/es_extended/tree/legacy

**Comment by yolobrandon01** [comment]
Okay dumb question on my part thanks, So legacy & v1 final is the same then? Sorry. I'm using the latest version of ESX v1, (v1 Final) And have been having hitch problems with esx_status for months and been trying to figure out a fix. Excited to see activity here on this thread. So downloading this should help with the server hitch issue then i take it.

**Comment by thelindat** [comment]
Basically once 1.2 stopped getting worked on it was released as v1 Final - that was just over a year ago now.
At this point the only thing to come to Legacy will be fixes or optimisations until it is finally archived. The same goes for many other resources such as this; so I'm resolving some of the issues with hitches.

There are no breaking changes being (intentionally) made so you can update from 1.2 or Final to Legacy without any issues. If something does break, just post an issue and I'll look into it. Also it's still a WIP but I am updating the readme to make things less confusing. https://github.com/thelindat/es_extended/tree/legacy#readme

**Comment by yolobrandon01** [comment]
Okay i'm going to try to update to es_extended Legacy (your version) and the pull request for the optimizations for esx_status (your pr) THANK YOU SO MUCH for working on ex_extended and esx_status legacy, I have been running a server on it for a year now and have much optimization issues :S

**Comment by thelindat** [comment]
Just make sure you get the other recently updated resources.

**Comment by yolobrandon01** [comment]
Everything seems to be working, Idk what I did wrong, but I can't get loadouts to load properly, It's just clearing the loadout upon connecting the server. It is saving the loadouts correctly though.

**Comment by yolobrandon01** [comment]
Okay so after a bit of testing, i'm using the ESX.GetExtendedPlayers to loop through and check ambulance job count.. it seemed to increase the hitch issue a tiny bit. 
![image](https://user-images.githubusercontent.com/66453444/121093745-758e3d80-c7bb-11eb-8fde-f20a3081b27f.png)

I may be because I'm just using your idea, i added the ESX.GetExtendedPlayers function to my version of es_extended, as well as changed the loop servercallback inside of the ambulance check to 

`ESX.RegisterServerCallback('esx_hospital:EmsCheck', function(source, cb)
--	local xPlayers = ESX.GetExtendedPlayers()
		local EmsConnected = 0
	for k, v in pairs(ESX.GetExtendedPlayers()) do
		local Player = ESX.GetPlayerFromId(v)
		if Player ~= nil then
			local EmsConnected = 0		
				if Player.job.name == 'ambulance' then
					EmsConnected = EmsConnected + 1
				end
			end	
		end				
	cb(EmsConnected)
--	print(EmsConnected)
end)`

The reason I just didnt download your version of es_extended (one i did and tried to convert it over and got it working but loadouts wouldnt save and two I'm using esx_kashacters so I have to edit your version of es_extended legacy to get it to work correcty.. i did that but again.. loadouts).. I am also using your version esx_status to try and get rid of the hitch problem.. Running into problems with this line of code for somereason though..

![image](https://user-images.githubusercontent.com/66453444/121094060-0107ce80-c7bc-11eb-8cb1-b67ae02b8357.png)

**Comment by ESXCrackhead** [comment]
To clarify something:

1.2 Final was a release, a release doesn't receive future updates after it's released, every update after a release is part a new version aka Legacy.

**Comment by yolobrandon01** [comment]
After updating to Legacy/Esx_Multicharacter/esx_identity and even linden_inventory. btw is a great resource nice work. I'm now optimizing and updating everything to work properly with it and its working out great, just wanted to update that status is still giving some high tick rates? i'll post a screenie, maybe something else I got going on is affect esx_status?

22 players online.. 
![image](https://user-images.githubusercontent.com/66453444/121618160-564e1500-ca34-11eb-912f-732b91fa7224.png)

**Comment by thelindat** [comment]
```lua
RegisterServerEvent('esx_status:update')
AddEventHandler('esx_status:update', function(status)
	local xPlayer = ESX.GetPlayerFromId(source)

	if xPlayer then
		xPlayer.set('status', status)
	end
end)
```

The major hitch was related to the status saving, this one is odd though - unless all your players joined back-to-back that event shouldn't be triggering enough in a short period to cause that much of a hitch.

Honestly I'm not sure why esx_status bothers setting player statuses in xPlayer when you never need to retrieve it anyway, unless I'm unaware of something. I'll bring it up to Taso.

**Comment by Tinky124** [comment]
Like i said https://github.com/esx-framework/esx_status/pull/53

Change `Config.UpdateInterval = 10000` to `Config.UpdateInterval = 60000` @yolobrandon01

**Comment by thelindat** [comment]
@Tinky124 @yolobrandon01 
> 22 players online..
> ![image](https://user-images.githubusercontent.com/66453444/121618160-564e1500-ca34-11eb-912f-732b91fa7224.png)

I believe I've found the source of the problem.
```lua
TriggerServerEvent('esx_status:update', GetStatusData(true))
```
Is already being triggered every x seconds (10 by default, which is honestly too often); however it also gets triggered by events.
`esx_status:set / esx_status:add / esx_status:remove`

Eating, drinking, or having these events triggered regularly will send the update event (source of the hitch as shown). If you have something like utk_stress then you are potentially triggering multiple stress events at once and sometimes overtime (per player).

Try disabling the triggerserverevent in the events I mentioned above.


You can also try this for less loops


```lua
RegisterNetEvent('esx_status:load')
AddEventHandler('esx_status:load', function(status)
	TriggerEvent('esx_status:loaded')
	for i=1, #Status, 1 do
		for j=1, #status, 1 do
			if Status[i].name == status[j].name then
				Status[i].set(status[j].val)
			end
		end
	end
	local updateTimer = 0
	Citizen.CreateThread(function()
		while #Status > 0 do
			for i=1, #Status, 1 do
				Status[i].onTick()
			end
			local data = GetStatusData(true)
			TriggerEvent('esx_status:onTick', data)
			if updateTimer == Config.UpdateInterval then TriggerServerEvent('esx_status:update', data) updateTimer = 0 end
			Citizen.Wait(Config.TickTime)
			updateTimer = updateTimer + Config.TickTime
		end
	end)
end)
```

Then at the bottom remove
```lua
Citizen.CreateThread(function()
	while true do
		Citizen.Wait(Config.UpdateInterval)

		TriggerServerEvent('esx_status:update', GetStatusData(true))
	end
end)
```
